File: src/api/subsonic/methods/savePlayQueue.ts
import { Fx } from '../../../_core/fx';
import {
  handleJsonResponse,
  makeSubsonicRequest,
  parseJsonResponse,
} from '../makeRequest';

export const subsonicSavePlayQueue = Fx.async(async function* f(
  id: string[],
  current?: string,
  { position = 0 }: { position?: number } = {},
) {
  const response = yield* makeSubsonicRequest({
    method: 'rest/savePlayQueue',
    current,
    id,
    position,
  });

  const json = yield* parseJsonResponse(response, {});

  return handleJsonResponse(json['subsonic-response']).map(() => {});
});

File: src/api/subsonic/savePlayQueue.ml
open Fx
open Monad_syntax
open Utils

let make id ?current ?(position = 0.0) () =
  let* { response = { status = `ok } } =
    let params = Js.Dict.empty () in
    Js.Dict.set params "id" (Array id);
    current
    |> Option.iter (fun current -> Js.Dict.set params "current" (Value current));
    Js.Dict.set params "position" (Value (position |> Js.Float.toString));
    makeRequest { _method = "rest/savePlayQueue"; params } Empty.of_json
  in
  Ok ()
